source/2_platform/rex_opengl/src/resources/texture2d.cpp

```diff
@@ -0,0 +1,191 @@
+                return GL_DEPTH_COMPONENT;
+
+            case TexturePixelFormat::DEPTH_STENCIL:
+                return GL_DEPTH_STENCIL;
+
+            case TexturePixelFormat::R:
+                return GL_RED;
+            case TexturePixelFormat::RG:
+                return GL_RG;
+            case TexturePixelFormat::RGB:
+                return GL_RGB;
+            case TexturePixelFormat::RGBA:
+                return GL_RGBA;
+
+            case TexturePixelFormat::BGR:
+                return GL_BGR;
+            case TexturePixelFormat::BGRA:
+                return GL_BGRA;
+            }
+
+            REX_ERROR("Unknown \"Texture Pixel Format\", returning \"GL_RGB\"");
+            return GL_RGB;
+        }
+
+        //-------------------------------------------------------------------------
+        GLenum get_texture_pixel_type(const DataType& type)
+        {
+            switch (type.get())
+            {
+            case DataType::Value::UNSIGNED_INT8:
+                return GL_UNSIGNED_BYTE;
+            }
+
+            REX_ERROR("Unknown \"Data Type\", returning \"GL_UNSIGNED_BYTE\"");
+            return GL_UNSIGNED_BYTE;
+        }
+    }
+
+    //-------------------------------------------------------------------------
+    Texture2D::Texture2D(const TextureDescription& desc, void* data)
+        :m_resource_id(0)
+        ,m_description(desc)
+    {
+        glGenTextures(1, &m_resource_id);
+        glBindTexture(GL_TEXTURE_2D, m_resource_id);
+
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, texture2d::get_texture_wrapping(m_description.s_wrap));
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, texture2d::get_texture_wrapping(m_description.t_wrap));
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, texture2d::get_texture_min_filter(m_description.filter));
+        glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, texture2d::get_texture_filter(m_description.filter));
+
+        REX_ASSERT(data != nullptr);
+
+        glTexImage2D(GL_TEXTURE_2D
+            , m_description.mipmap_level
+            , texture2d::get_texture_internal_format(m_description.storage_format)
+            , m_description.width
+            , m_description.height
+            , 0
+            , texture2d::get_texture_pixel_format(m_description.pixel_format)
+            , texture2d::get_texture_pixel_type(m_description.pixel_data_type)
+            , data);
+
+        glGenerateMipmap(GL_TEXTURE_2D);
+    }
+
+    //-------------------------------------------------------------------------
+    Texture2D::~Texture2D()
+    {
+        release();
+    }
+
+    //-------------------------------------------------------------------------
+    u32 Texture2D::get_resource_id() const
+    {
+        return m_resource_id;
+    }
+
+    //-------------------------------------------------------------------------
+    void Texture2D::bind() const
+    {
+        glBindTexture(GL_TEXTURE_2D, m_resource_id);
```

Comment:
put an assert here to make sure we're binding a valid texture (m_resource_id != 0)